### DISCUSSION TITLE
Executing bulk of short commands in queue

### DISCUSSION THREAD

**Discussion by DanielSpravtsev** [phpseclib/phpseclib#1927] [Q&A] [answered]
Hey there,

_Is there a way to decrease time to login process?_ 
I am **using queues** in my application, so I need to do $ssh->login in every Job class?
As I understand, there is no way to save login session / cache or serialize ssh connection to next callable php job via queue?

In this case, I think there is only one way to fasten the authentication — using ssh-agent or pubkey auth, is it right? 

@terrafrost Will be glad for your tips ❤️

**Comment by terrafrost** [comment, marked as answer]
ssh-agent wouldn't really help that much. All that really does is give you a way to create a signature with a private key and retrieve the public key corresponding to a private key without being able to actually read that private key. The signatures and key exchange and all that jazz still needs to be done.

Unless you could make it so that the server doesn't randomly generate numbers there's not a whole lot you can do to cache logins. And if you did make it so that the server served up the same random bytes every time then you'd be introducing security vulnerabilities.

I think the best option, in your situation, is going to be to just make it so phpseclib is always running. Do something like this:

```php
$ssh = new SSH2(...);
$ssh->login(...);
while (true) {
    // query MySQL DB for commands or some such
    $commands = [];
    foreach ($commands as $command) {
        $result = $ssh->exec($command);
        // save command output in DB
    }
    sleep(1);
}

**Reply by DanielSpravtsev** [reply]
I'm think I just get rid of 1 job = 1 exec and just run it in one parent job. Thanks for advice, much appreciated.
